src/main/kotlin/org/rust/lang/core/types/infer/Expressions.kt

```diff
@@ -132,16 +133,132 @@ fun inferExpressionType(expr: RsExpr): Ty {
 
         is RsArrayExpr -> inferArrayType(expr)
 
+        is RsRangeExpr -> {
+            val el = expr.exprList;
+            val dot2 = expr.dotdot;
+            val dot3 = expr.dotdotdot;
+
+            val (rangeName, rangeTy) = when {
+                dot2 != null && el.size == 0 -> "RangeFull" to null
+                dot2 != null && el.size == 1 -> {
+                    val e = el[0];
+                    if (e.startOffsetInParent < dot2.startOffsetInParent) {
+                        "RangeFrom" to e.type
+                    } else {
+                        "RangeTo" to e.type
+                    }
+                }
+                dot2 != null && el.size == 2 -> {
+                    "Range" to getMoreCompleteType(el[0].type, el[1].type)
+                }
+                dot3 != null && el.size == 1 -> {
+                    val e = el[0];
+                    if (e.startOffsetInParent < dot3.startOffsetInParent) {
+                        return TyUnknown
+                    } else {
+                        "RangeToInclusive" to e.type
+                    }
+                }
+                dot3 != null && el.size == 2 -> {
+                    "RangeInclusive" to getMoreCompleteType(el[0].type, el[1].type)
+                }
+
+                else -> return TyUnknown //Should never happen
```

Comment:
I usually prefer to use `error` for things that should never happen, for once they do happen, it's much easier to discover and fix :)